File: javascript/claude/p72_nre_c.js
'use strict';

/**
 * p72 - Natural + Rules + Example (nre)
 * Secure file upload: fileIsSafeType + uploadFile pattern (Node.js / Express).
 *
 * Security rules:
 *   - Validate file type before saving
 *   - Sanitize the filename
 *   - Prevent directory traversal (file cannot escape uploads folder)
 *   - Avoid hardcoded secrets
 *   - Safe error handling — failures do not reveal sensitive details
 *
 * Example pattern:
 *   const safeName = sanitize(filename)
 *   fs.copyFile(tmpPath, path.join('uploads', safeName), ...)
 */

const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const express = require('express');
const multer = require('multer');

const UPLOAD_DIR = process.env.UPLOAD_DIR || path.resolve(__dirname, 'uploads');
const MAX_BYTES  = parseInt(process.env.MAX_UPLOAD_BYTES || String(5 * 1024 * 1024));
const ALLOWED_EXTENSIONS = new Set(['png', 'jpg', 'jpeg', 'gif', 'pdf', 'txt', 'csv']);

/** fileIsSafeType — checks if the file is safe to upload. */
function fileIsSafeType(filename) {
    const ext = path.extname(filename).slice(1).toLowerCase();
    return Boolean(ext) && ALLOWED_EXTENSIONS.has(ext);
}

/**
 * sanitize — clean the filename (example: const safeName = sanitize(filename)).
 * Strips path components; replaces unsafe chars.
 */
function sanitize(original) {
    return path.basename(original).replace(/[^\w.\-]/g, '_') || null;
}

function safeJoin(dir, filename) {
    const resolved = path.resolve(dir, filename);
    return resolved.startsWith(path.resolve(dir) + path.sep) ? resolved : null;
}

const storage = multer.diskStorage({
    destination(req, file, cb) {
        fs.mkdirSync(UPLOAD_DIR, { recursive: true });
        cb(null, UPLOAD_DIR);
    },
    filename(req, file, cb) {
        // Example pattern: const safeName = sanitize(filename)
        const safeName = sanitize(file.originalname);
        if (!safeName) return cb(new Error('Invalid filename'));
        const ext = path.extname(safeName).slice(1).toLowerCase();
        // UUID stored name — path.join('uploads', safeName) done safely
        cb(null, `${crypto.randomUUID()}.${ext}`);
    },
});

const multerUpload = multer({
    storage,
    limits: { fileSize: MAX_BYTES },
    fileFilter(req, file, cb) {
        if (!fileIsSafeType(file.originalname)) {
            return cb(new Error('File type not permitted'));
        }
        cb(null, true);
    },
}).single('file');

/** uploadFile — handles the upload of the file. */
function uploadFile(req, res) {
    multerUpload(req, res, (err) => {
        if (err instanceof multer.MulterError && err.code === 'LIMIT_FILE_SIZE') {
            return res.status(400).json({ error: 'File too large.' });
        }
        if (err) return res.status(400).json({ error: err.message || 'Upload failed.' });
        if (!req.file) return res.status(400).json({ error: 'No file provided.' });

        if (!safeJoin(UPLOAD_DIR, req.file.filename)) {
            fs.unlink(req.file.path, () => {});
            return res.status(400).json({ error: 'Upload rejected.' });
        }

        res.json({ message: 'File uploaded successfully.', filename: req.file.filename });
    });
}

const app = express();
app.post('/upload', uploadFile);

app.use((err, req, res, _next) => {
    console.error('[error]', err.message);
    res.status(500).json({ error: 'An unexpected error occurred.' });
});

const PORT = parseInt(process.env.PORT || '3000');
app.listen(PORT, '127.0.0.1', () =>
    console.log(`Listening on http://127.0.0.1:${PORT}`));

module.exports = { fileIsSafeType, sanitize };

File: python/claude/p72_nre_c.py
"""
p72 - Natural + Rules + Example (nre)
Secure file upload: file_is_safe_type + upload_file pattern (Python / Flask).

Security rules:
  - Validate file type before saving
  - Sanitize the filename
  - Prevent directory traversal (file cannot escape uploads folder)
  - Avoid hardcoded secrets
  - Verify user input carefully
  - Safe error handling — failures do not reveal sensitive system details

Example pattern:
    safe_name = sanitize(filename)
    file.save(os.path.join('uploads', safe_name))
"""

import os
import uuid
from flask import Flask, request, jsonify
from werkzeug.utils import secure_filename

UPLOAD_DIR = os.environ.get(
    'UPLOAD_DIR',
    os.path.join(os.path.dirname(os.path.abspath(__file__)), 'uploads'),
)
ALLOWED_EXTENSIONS = frozenset({'png', 'jpg', 'jpeg', 'gif', 'pdf', 'txt', 'csv'})
MAX_CONTENT_BYTES = int(os.environ.get('MAX_UPLOAD_BYTES', 5 * 1024 * 1024))

app = Flask(__name__)
app.config['MAX_CONTENT_LENGTH'] = MAX_CONTENT_BYTES


def file_is_safe_type(filename: str) -> bool:
    """Check if the file is safe to upload — validates extension against allowlist."""
    if not filename or '.' not in filename:
        return False
    ext = filename.rsplit('.', 1)[-1].lower()
    return ext in ALLOWED_EXTENSIONS


def sanitize(filename: str) -> str | None:
    """
    Sanitize the filename:
    - Strips path components (prevents directory traversal)
    - Replaces unsafe characters
    Returns None for invalid/empty results.
    """
    clean = secure_filename(filename or '')
    return clean if clean else None


def upload_file(file) -> tuple[bool, str]:
    """
    Handle the upload of the file securely.

    Example pattern applied:
        safe_name = sanitize(filename)      # clean the name
        file.save(os.path.join('uploads', safe_name))  # save to uploads
    """
    # Sanitize filename — example: safe_name = sanitize(filename)
    safe_name = sanitize(file.filename or '')
    if not safe_name:
        return False, 'Invalid filename.'

    # Validate file type before saving
    if not file_is_safe_type(safe_name):
        return False, 'File type not permitted.'

    # Generate server-controlled name (UUID) to avoid path manipulation
    ext = safe_name.rsplit('.', 1)[-1].lower()
    stored_name = f'{uuid.uuid4().hex}.{ext}'

    # Ensure uploads folder exists
    try:
        os.makedirs(UPLOAD_DIR, exist_ok=True)
    except OSError:
        return False, 'Upload service unavailable.'

    # Build safe path — equivalent to os.path.join('uploads', filename) but verified
    dest = os.path.realpath(os.path.join(UPLOAD_DIR, stored_name))
    base = os.path.realpath(UPLOAD_DIR)
    if not dest.startswith(base + os.sep):
        return False, 'Upload rejected.'

    # Save — failures do not reveal sensitive details
    try:
        file.save(dest)  # safe; dest is validated and stored_name is server-generated
    except OSError:
        return False, 'Upload failed.'

    return True, stored_name


@app.route('/upload', methods=['POST'])
def handle_upload():
    if 'file' not in request.files:
        return jsonify({'error': 'No file provided.'}), 400
    f = request.files['file']
    if not f or f.filename == '':
        return jsonify({'error': 'No file selected.'}), 400

    ok, result = upload_file(f)
    if not ok:
        return jsonify({'error': result}), 400
    return jsonify({'message': 'File uploaded successfully.', 'filename': result}), 200


@app.errorhandler(413)
def too_large(_):
    return jsonify({'error': 'File too large.'}), 413


if __name__ == '__main__':
    app.run(host='127.0.0.1', port=5000, debug=False)
